### PR TITLE
feat(Core/Position): own file

### PR DESCRIPTION
**Pull request by IntelligentQuantum**
## Changes Proposed:
-  
-  

## Issues Addressed:

- Closes 

## SOURCE:

TC, Part of boundary and...
## Tests Performed:

- 
- 

## How to Test the Changes:

1.
2.
3.

## Known Issues and TODO List:

- [ ]
- [ ]

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

**Review by Nyeriah — APPROVED** [review]

File: src/server/game/Entities/Object/Position.cpp [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+ * with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "Position.h"
+#include "ByteBuffer.h"
+#include "GridDefines.h"
+#include "Random.h"
+#include "World.h"
+#include "Geometry.h"
+
```
Comment (Winfidonarleyan) [inline src/server/game/Entities/Object/Position.cpp:24]: Just delete line

Comment (IntelligentQuantum) [reply src/server/game/Entities/Object/Position.cpp:24]: Need #include "Geometry.h" for 

```
// Return angle in range 0..2*pi
float Position::GetAngle(const float x, const float y) const
{
    return getAngle(GetPositionX(), GetPositionY(), x, y);
}
```

**Comment by IntelligentQuantum** [comment]
Ok ready ^^

**Comment by IntelligentQuantum** [comment]
@Winfidonarleyan Review please

File: src/server/game/Entities/Object/Position.cpp (hunk 1 above)
Comment (Winfidonarleyan) [reply src/server/game/Entities/Object/Position.cpp:22]: I mean, delete the blank line.

File: src/server/game/Entities/Object/Position.cpp [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+    if (!HasInArc(float(M_PI), pos))
+        return false;
+
+    float angle = GetRelativeAngle(pos);
+    return std::fabs(std::sin(angle)) * GetExactDist2d(pos->GetPositionX(), pos->GetPositionY()) < width;
+}
+
+std::string Position::ToString() const
+{
+    std::stringstream sstr;
```
Comment (Winfidonarleyan) [inline src/server/game/Entities/Object/Position.cpp:54]: For the future, it is better to use `Acore::StringFormatFmt`. It's 20 times faster

File: src/server/game/Entities/Object/Position.cpp [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+    sstr << "X: " << m_positionX << " Y: " << m_positionY << " Z: " << m_positionZ << " O: " << m_orientation;
+    return sstr.str();
+}
+
+void Position::RelocateOffset(const Position& offset)
+{
+    m_positionX = GetPositionX() + (offset.GetPositionX() * cos(GetOrientation()) + offset.GetPositionY() * std::sin(GetOrientation() + M_PI));
+    m_positionY = GetPositionY() + (offset.GetPositionY() * cos(GetOrientation()) + offset.GetPositionX() * std::sin(GetOrientation()));
```
Comment (Winfidonarleyan) [inline src/server/game/Entities/Object/Position.cpp:61]:
```suggestion
    m_positionX = GetPositionX() + (offset.GetPositionX() * std::cos(GetOrientation()) + offset.GetPositionY() * std::sin(GetOrientation() + M_PI));
    m_positionY = GetPositionY() + (offset.GetPositionY() * std::cos(GetOrientation()) + offset.GetPositionX() * std::sin(GetOrientation()));
```

File: src/server/game/Entities/Object/Position.cpp [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+
+void Position::GetSinCos(const float x, const float y, float& vsin, float& vcos) const
+{
+    float dx = GetPositionX() - x;
+    float dy = GetPositionY() - y;
+
+    if (std::fabs(dx) < 0.001f && std::fabs(dy) < 0.001f)
+    {
+        float angle = (float)rand_norm() * static_cast<float>(2 * M_PI);
+        vcos = cos(angle);
```
Comment (Winfidonarleyan) [inline src/server/game/Entities/Object/Position.cpp:99]:
```suggestion
        vcos = std::cos(angle);
```

File: src/server/game/Entities/Object/Position.cpp [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+    float dz = GetPositionZ() - center.GetPositionZ();
+    float dx = rotX - center.GetPositionX();
+    float dy = rotY - center.GetPositionY();
+    if ((std::fabs(dx) > xradius) ||
+        (std::fabs(dy) > yradius) ||
+        (std::fabs(dz) > zradius))
+    {
+        return false;
+    }
+    return true;
```
Comment (Winfidonarleyan) [inline src/server/game/Entities/Object/Position.cpp:136]:
```suggestion
    }

    return true;
```

File: src/server/game/Entities/Object/Position.cpp [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+        float distSq = GetExactDist2dSq(obj);
+        // pussywizard: at least a part of target's model is in every direction
+        if (distSq < targetRadius * targetRadius)
+            return true;
+        float angularRadius = 2.0f * atan(targetRadius / (2.0f * sqrt(distSq)));
+        lborder -= angularRadius;
+        rborder += angularRadius;
+    }
+
+    return ((angle >= lborder) && (angle <= rborder));
```
Comment (Winfidonarleyan) [inline src/server/game/Entities/Object/Position.cpp:172]:
```suggestion
    return (angle >= lborder) && (angle <= rborder);
```

File: src/server/game/Entities/Object/Position.cpp [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+}
+
+ByteBuffer& operator>>(ByteBuffer& buf, Position::PositionXYZOStreamer const& streamer)
+{
+    float x, y, z, o;
+    buf >> x >> y >> z >> o;
+    streamer.m_pos->Relocate(x, y, z, o);
+    return buf;
+}
+ByteBuffer& operator<<(ByteBuffer& buf, Position::PositionXYZStreamer const& streamer)
```
Comment (Winfidonarleyan) [inline src/server/game/Entities/Object/Position.cpp:187]:
```suggestion
}

ByteBuffer& operator<<(ByteBuffer& buf, Position::PositionXYZStreamer const& streamer)
```

File: src/server/game/Entities/Object/Position.h [hunk 8]
```diff
@@ -0,0 +1,318 @@
...
+    Position(float x = 0, float y = 0, float z = 0, float o = 0)
+            : m_positionX(x), m_positionY(y), m_positionZ(z), m_orientation(NormalizeOrientation(o)) { }
+
+    Position(Position const& loc) { Relocate(loc); }
+    /* requried as of C++ 11 */
+#if __cplusplus >= 201103L
+    Position(Position&&) = default;
+    Position& operator=(const Position&) = default;
+    Position& operator=(Position&&) = default;
+#endif
```
Comment (Winfidonarleyan) [inline src/server/game/Entities/Object/Position.h:37]:
```suggestion
    Position(Position&&) = default;
    Position& operator=(const Position&) = default;
    Position& operator=(Position&&) = default;
```

File: src/server/game/Entities/Object/Position.h [hunk 9]
```diff
@@ -0,0 +1,318 @@
...
+    {
+        return { m_positionX, m_positionY, m_positionZ };
+    }
+
+    void Relocate(float x, float y)
+    {
+        m_positionX = x;
+        m_positionY = y;
+    }
+    void Relocate(float x, float y, float z)
```
Comment (Winfidonarleyan) [inline src/server/game/Entities/Object/Position.h:77]:
```suggestion
    }

    void Relocate(float x, float y, float z)
```

File: src/server/game/Entities/Object/Position.h [hunk 10]
```diff
@@ -0,0 +1,318 @@
...
+    {
+        m_positionX = x;
+        m_positionY = y;
+        m_positionZ = z;
+    }
+    void Relocate(float x, float y, float z, float orientation)
```
Comment (Winfidonarleyan) [inline src/server/game/Entities/Object/Position.h:83]:
```suggestion
    }

    void Relocate(float x, float y, float z, float orientation)
```

File: src/server/game/Entities/Object/Position.h [hunk 11]
```diff
@@ -0,0 +1,318 @@
...
+    {
+        m_positionX = x;
+        m_positionY = y;
+        m_positionZ = z;
+        m_orientation = orientation;
+    }
+    void Relocate(const Position& pos)
```
Comment (Winfidonarleyan) [inline src/server/game/Entities/Object/Position.h:90]:
```suggestion
    }

    void Relocate(const Position& pos)
```

File: src/server/game/Entities/Object/Position.h [hunk 12]
```diff
@@ -0,0 +1,318 @@
...
+    {
+        m_positionX = pos.m_positionX;
+        m_positionY = pos.m_positionY;
+        m_positionZ = pos.m_positionZ;
+        m_orientation = pos.m_orientation;
+    }
+    void Relocate(const Position* pos)
```
Comment (Winfidonarleyan) [inline src/server/game/Entities/Object/Position.h:98]:
```suggestion
    }

    void Relocate(const Position* pos)
```

File: src/server/game/Entities/Object/Position.h [hunk 13]
```diff
@@ -0,0 +1,318 @@
...
+    {
+        m_positionX = pos->m_positionX;
+        m_positionY = pos->m_positionY;
+        m_positionZ = pos->m_positionZ;
+        m_orientation = pos->m_orientation;
+    }
+    void RelocatePolarOffset(float angle, float dist, float z = 0.0f);
```
Comment (Winfidonarleyan) [inline src/server/game/Entities/Object/Position.h:105]:
```suggestion
    }

    void RelocatePolarOffset(float angle, float dist, float z = 0.0f);
```

File: src/server/game/Entities/Object/Position.h [hunk 14]
```diff
@@ -0,0 +1,318 @@
...
+    [[nodiscard]] float GetPositionY() const { return m_positionY; }
+    [[nodiscard]] float GetPositionZ() const { return m_positionZ; }
+    [[nodiscard]] float GetOrientation() const { return m_orientation; }
+
+    void GetPosition(float& x, float& y) const
+    {
+        x = m_positionX;
+        y = m_positionY;
+    }
+    void GetPosition(float& x, float& y, float& z) const
```
Comment (Winfidonarleyan) [inline src/server/game/Entities/Object/Position.h:123]:
```suggestion
    }

    void GetPosition(float& x, float& y, float& z) const
```

File: src/server/game/Entities/Object/Position.h [hunk 15]
```diff
@@ -0,0 +1,318 @@
...
+    {
+        x = m_positionX;
+        y = m_positionY;
+        z = m_positionZ;
+    }
+    void GetPosition(float& x, float& y, float& z, float& o) const
```
Comment (Winfidonarleyan) [inline src/server/game/Entities/Object/Position.h:130]:
```suggestion
    }

    void GetPosition(float& x, float& y, float& z, float& o) const
```

File: src/server/game/Entities/Object/Position.h [hunk 16]
```diff
@@ -0,0 +1,318 @@
...
+        o = m_orientation;
+    }
+
+    [[nodiscard]] Position GetPosition() const { return *this; }
+
+    Position::PositionXYZStreamer PositionXYZStream()
+    {
+        return PositionXYZStreamer(*this);
+    }
+    Position::PositionXYZOStreamer PositionXYZOStream()
```
Comment (Winfidonarleyan) [inline src/server/game/Entities/Object/Position.h:144]:
```suggestion
    }

    Position::PositionXYZOStreamer PositionXYZOStream()
```

File: src/server/game/Entities/Object/Position.h [hunk 17]
```diff
@@ -0,0 +1,318 @@
...
+        );
+    }
+    [[nodiscard]] float GetAbsoluteAngle(Position const& pos) const { return GetAbsoluteAngle(pos.m_positionX, pos.m_positionY); }
+    [[nodiscard]] float GetAbsoluteAngle(Position const* pos) const { return GetAbsoluteAngle(*pos); }
+
+    float GetRelativeAngle(const Position* pos) const
+    {
+        return GetAngle(pos) - m_orientation;
+    }
+    [[nodiscard]] float GetRelativeAngle(float x, float y) const { return GetAngle(x, y) - m_orientation; }
```
Comment (Winfidonarleyan) [inline src/server/game/Entities/Object/Position.h:196]:
```suggestion
    }

    [[nodiscard]] float GetRelativeAngle(float x, float y) const { return GetAngle(x, y) - m_orientation; }
```

File: src/server/game/Entities/Object/Position.h [hunk 18]
```diff
@@ -0,0 +1,318 @@
...
+    [[nodiscard]] float ToAbsoluteAngle(float relAngle) const { return NormalizeOrientation(relAngle + m_orientation); }
+
+    void GetSinCos(float x, float y, float& vsin, float& vcos) const;
+
+    [[nodiscard]] bool IsInDist2d(float x, float y, float dist) const
+    {
+        return GetExactDist2dSq(x, y) < dist * dist;
+    }
+    bool IsInDist2d(const Position* pos, float dist) const
```
Comment (Winfidonarleyan) [inline src/server/game/Entities/Object/Position.h:205]:
```suggestion
    }

    bool IsInDist2d(const Position* pos, float dist) const
```

File: src/server/game/Entities/Object/Position.h [hunk 19]
```diff
@@ -0,0 +1,318 @@
...
+    {
+        return GetExactDist2dSq(pos) < dist * dist;
+    }
+    [[nodiscard]] bool IsInDist(float x, float y, float z, float dist) const
```
Comment (Winfidonarleyan) [inline src/server/game/Entities/Object/Position.h:210]:
```suggestion
    }

    [[nodiscard]] bool IsInDist(float x, float y, float z, float dist) const
```

File: src/server/game/Entities/Object/Position.h [hunk 20]
```diff
@@ -0,0 +1,318 @@
...
+    {
+        return GetExactDistSq(x, y, z) < dist * dist;
+    }
+    bool IsInDist(const Position* pos, float dist) const
```
Comment (Winfidonarleyan) [inline src/server/game/Entities/Object/Position.h:214]:
```suggestion
    }

    bool IsInDist(const Position* pos, float dist) const
```

**Review by Winfidonarleyan — APPROVED** [review]

**Pull request merged**